base/libgit2/types.jl

```diff
@@ -612,13 +631,54 @@ function getobjecttype(obj_type::Cint)
     end
 end
 
-type UserPasswordCredentials <: AbstractPayload
+"Empty credentials"
+type EmptyCredentials <: AbstractCredentials end
+
+"Credentials that support only `user` and `password` parameters"
+type UserPasswordCredentials <: AbstractCredentials
     user::AbstractString
     pass::AbstractString
-    used::Bool
-    UserPasswordCredentials(u::AbstractString,p::AbstractString) = new(u,p,false)
+    count::Int
+    use_ssh_agent::UInt8
+    UserPasswordCredentials(u::AbstractString,p::AbstractString) = new(u,p,3,zero(UInt8))
+end
+function isused(p::UserPasswordCredentials)
+    p.count <= 0 && return true
+    p.count -= 1
```

Comment:
This is because of the specifics of libgit2 authentication procedure: when authentication is failed, libgit2 calls credentials callback again, but there is no information whether authentication was successful or not. So, in order not to stuck in infinite loop by using same faulty credentials simple counting strategy is implemented.